### PR TITLE
Refactor UART and allow UART via DMA (not enabled by default yet)

### PR DESCRIPTION
**Pull request by asmodai27**
# Description
Refactoring of UART and Comms:
- make comms a class rather than a collection of static stuff
- remove UART registers from h file
- communalize UART and UART_DMA interface
- allow better mocking of UART
- define UART channels independently from hal, and remove hal accessors to UART

Tested on my ventilator (pizza s/n 7) with both default (UART handled by SW) and `stm32-uart` environments. We could at this point remove stm32-uart and get rid of all the `#if !defined(UART_VIA_DMA)` content, though I would prefer another tester to have a look at this before going forward.
Update from @inceptionev's tests: synchronization of the GUI->controller data seems not to work properly when using `UART_VIA_DMA`, which seems due to the fact that if the GUI starts sending data before the `comms` module has initiated a transfer (by calling `uart_->StartRx([...])`), the beginning of the data is lost to the controller, but DMA always handles `GuiStatusSize` bytes, leading to composite messages being sent to comms, which cannot de-serialize them.

Updates #1184, leaving it with only stepper/SPI to disentangle.

TODO:
- [x] address cppcheck and clangtidy warnings.
- [ ] address DMA issue (framing PR, in progress)

# General checklist:

- [x] I have performed a self-review, including - looked through the `Files changed` tab, browsed repository in my branch
- [ ] I have made corresponding changes to the documentation - to reflect changes in code, electrical or mechanical design
- [x] All new documentation or graphics follow the [documentation style guide](https://github.com/RespiraWorks/Ventilator/wiki/Documentation-style-guide)
- [x] All new content is linked, easily discoverable, does not require too many clicks
- [x] I have reviewed any other relevant parts of our [contributor wiki](https://github.com/RespiraWorks/Ventilator/wiki) to ensure that this PR conforms to the standards
- [x] I have given the PR a descriptive name (prefixed with **WIP** if it is not yet ready for review)
- [x] I have tagged relevant reviewers

## Code checklist:

- [x] All new code follows our [code style](https://github.com/RespiraWorks/Ventilator/wiki/Code-style)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have described tests I used to verify my changes, instructions for reproducing them and any relevant configuration details.
- [x] I ran the unit test suite and verified that all tests pass - new and old, locally and on CI
- [x] I performed a clean build and verified that there were no warnings
- [x] I ran our static analysis tools and verified there were no warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] All source files have license headers

### CODE REVIEW

**Comment by mortenjc** [comment]
To whom it may concern - I've been invited to do a code review on this PR. I will do my best and try to provide some 
comments. Since I really do not know the architecture that well and certainly not the microcontrollers, etc. my comments 
will very likely be a random selection of stuff I have observed. Feel free to ignore. I will start looking at this today.

**Review by mortenjc — APPROVED** [review]
There is some nice refactoring in here and I have posted some minor comments. A few general remarks that covers the entire project:

1) asserts
I would recommend investigating using assert() (using assert.h or using your own implementation) to make sure that you fail fast when not in release mode

If interested read this article on assert and friends to get inspiration
https://queue.acm.org/detail.cfm?id=2220317

2) internal stat counters
generally I would recommend to put some effort into using realtime counters
for error conditions, warnings, cases that (probably) happens rarely, etc.
These could be 32 bit counters in a single struct or - depending on your threading model - something slightly more complicated.
But in any case exposing these internal counters by some method (probably debug commands or periodic reporting) might  be valuable in building up an understanding of system level behaviour.
I mentioned in one of the comments that I could give a demo of how I use this.

3) TODO - Doxygen
If you use the Doxygen code documentation standard you can use the keyword \todo 
and when running Doxygen you will get a htmp page with (among other nice things)
a compiled list of all the TODOs in the project.

So nice to see unit tests and as always - keep writing more!

I am afraid that this was all I could contribute with. Thanks for the invitation. I will 'approve' but this just means that I should not block the further progress on this PR.

Cheers
M

File: software/controller/lib/core/comms.cpp [hunk 1]
```diff
@@ -69,69 +54,78 @@ static void ProcessTx(const ControllerStatus &controller_status) {
   //  - we're not currently transmitting,
```
[the remark this answers is not in the document]
Comment (asmodai27) [reply software/controller/lib/core/comms.cpp:54]: Will refine the comments, not sure this mandates a separate function if the comments are made clearer.

**Review by asmodai27 — COMMENTED** [review]
I won't merge yet because of issues with `platformio` which makes tests fail on CI, but I have addressed those comments.

**Pull request merged**